lib/transaction.js

```diff
@@ -195,7 +195,7 @@ class Transaction extends EventEmitter {
 
             value = new Error(`Transaction rejected with non-error: ${value}`);
           }
-          this._rejecter(value);
+          setImmediate(() => this._rejecter(value));
```

Comment:
I just tried out your fix, but I'm still observing the problem within this test case:

https://github.com/knex/knex/blob/77df705189c712105cfab293ae26fff41de8e6ae/test/integration/builder/additional.js#L863

Here's a sample of my output:

```
 % DB="mysql mysql2" KNEX_TEST_TIMEOUT=60000 ./node_modules/.bin/mocha --exit -t 10000 test/index.js
sqlite does not support inserting default values. Set the `useNullAsDefault` flag to hide this warning. (see docs http://knexjs.org/#Builder-insert).
sqlite does not support inserting default values. Set the `useNullAsDefault` flag to hide this warning. (see docs http://knexjs.org/#Builder-insert).
sqlite does not support inserting default values. Set the `useNullAsDefault` flag to hide this warning. (see docs http://knexjs.org/#Builder-insert).


  Integration Tests
    mysql | mysql
      Additional
Unhandled rejection: KnexTimeoutError: Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?
    at /Users/brianlauber/Documents/src/knex/lib/client.js:351:30
From previous event:
    at Client_MySQL.acquireConnection (/Users/brianlauber/Documents/src/knex/lib/client.js:348:15)
    at Client_MySQL.cancelQuery (/Users/brianlauber/Documents/src/knex/lib/dialects/mysql/index.js:175:32)
    at /Users/brianlauber/Documents/src/knex/lib/runner.js:177:37
From previous event:
    at Runner.query (/Users/brianlauber/Documents/src/knex/lib/runner.js:169:13)
    at /Users/brianlauber/Documents/src/knex/lib/runner.js:41:23
    at /Users/brianlauber/Documents/src/knex/lib/runner.js:261:24
From previous event:
    at Runner.ensureConnection (/Users/brianlauber/Documents/src/knex/lib/runner.js:259:8)
    at Runner.run (/Users/brianlauber/Documents/src/knex/lib/runner.js:27:12)
    at Raw.Target.then (/Users/brianlauber/Documents/src/knex/lib/interface.js:15:43)
    at Context.<anonymous> (/Users/brianlauber/Documents/src/knex/test/integration/builder/additional.js:908:10)
    at callFn (/Users/brianlauber/Documents/src/knex/node_modules/mocha/lib/runnable.js:395:21)
    at Test.Runnable.run (/Users/brianlauber/Documents/src/knex/node_modules/mocha/lib/runnable.js:382:7)
    at Runner.runTest (/Users/brianlauber/Documents/src/knex/node_modules/mocha/lib/runner.js:541:10)
    at /Users/brianlauber/Documents/src/knex/node_modules/mocha/lib/runner.js:667:12
    at next (/Users/brianlauber/Documents/src/knex/node_modules/mocha/lib/runner.js:450:14)
    at /Users/brianlauber/Documents/src/knex/node_modules/mocha/lib/runner.js:460:7
    at next (/Users/brianlauber/Documents/src/knex/node_modules/mocha/lib/runner.js:362:14)
    at Immediate.<anonymous> (/Users/brianlauber/Documents/src/knex/node_modules/mocha/lib/runner.js:428:5)
    at processImmediate (internal/timers.js:439:21) {
  name: 'KnexTimeoutError'
}
        ✓ .timeout(ms, {cancel: true}) should throw error if cancellation cannot acquire connection (5135ms)
    mysql | mysql2
      Additional
Unhandled rejection: KnexTimeoutError: Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?
    at /Users/brianlauber/Documents/src/knex/lib/client.js:351:30
From previous event:
    at Client_MySQL2.acquireConnection (/Users/brianlauber/Documents/src/knex/lib/client.js:348:15)
    at Client_MySQL2.cancelQuery (/Users/brianlauber/Documents/src/knex/lib/dialects/mysql/index.js:175:32)
    at /Users/brianlauber/Documents/src/knex/lib/runner.js:177:37
From previous event:
    at Runner.query (/Users/brianlauber/Documents/src/knex/lib/runner.js:169:13)
    at /Users/brianlauber/Documents/src/knex/lib/runner.js:41:23
    at /Users/brianlauber/Documents/src/knex/lib/runner.js:261:24
From previous event:
    at Runner.ensureConnection (/Users/brianlauber/Documents/src/knex/lib/runner.js:259:8)
    at Runner.run (/Users/brianlauber/Documents/src/knex/lib/runner.js:27:12)
    at Raw.Target.then (/Users/brianlauber/Documents/src/knex/lib/interface.js:15:43)
    at Context.<anonymous> (/Users/brianlauber/Documents/src/knex/test/integration/builder/additional.js:908:10)
    at callFn (/Users/brianlauber/Documents/src/knex/node_modules/mocha/lib/runnable.js:395:21)
    at Test.Runnable.run (/Users/brianlauber/Documents/src/knex/node_modules/mocha/lib/runnable.js:382:7)
    at Runner.runTest (/Users/brianlauber/Documents/src/knex/node_modules/mocha/lib/runner.js:541:10)
    at /Users/brianlauber/Documents/src/knex/node_modules/mocha/lib/runner.js:667:12
    at next (/Users/brianlauber/Documents/src/knex/node_modules/mocha/lib/runner.js:450:14)
    at /Users/brianlauber/Documents/src/knex/node_modules/mocha/lib/runner.js:460:7
    at next (/Users/brianlauber/Documents/src/knex/node_modules/mocha/lib/runner.js:362:14)
    at Immediate.<anonymous> (/Users/brianlauber/Documents/src/knex/node_modules/mocha/lib/runner.js:428:5)
    at processImmediate (internal/timers.js:439:21) {
  name: 'KnexTimeoutError'
}
        ✓ .timeout(ms, {cancel: true}) should throw error if cancellation cannot acquire connection (5125ms)


  2 passing (10s)

Unhandled rejections: 2
{
  reason: KnexTimeoutError: Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?
      at /Users/brianlauber/Documents/src/knex/lib/client.js:351:30
  From previous event:
      at Client_MySQL.acquireConnection (/Users/brianlauber/Documents/src/knex/lib/client.js:348:15)
      at Client_MySQL.cancelQuery (/Users/brianlauber/Documents/src/knex/lib/dialects/mysql/index.js:175:32)
      at /Users/brianlauber/Documents/src/knex/lib/runner.js:177:37
  From previous event:
      at Runner.query (/Users/brianlauber/Documents/src/knex/lib/runner.js:169:13)
      at /Users/brianlauber/Documents/src/knex/lib/runner.js:41:23
      at /Users/brianlauber/Documents/src/knex/lib/runner.js:261:24
  From previous event:
      at Runner.ensureConnection (/Users/brianlauber/Documents/src/knex/lib/runner.js:259:8)
      at Runner.run (/Users/brianlauber/Documents/src/knex/lib/runner.js:27:12)
      at Raw.Target.then (/Users/brianlauber/Documents/src/knex/lib/interface.js:15:43)
      at Context.<anonymous> (/Users/brianlauber/Documents/src/knex/test/integration/builder/additional.js:908:10)
      at callFn (/Users/brianlauber/Documents/src/knex/node_modules/mocha/lib/runnable.js:395:21)
      at Test.Runnable.run (/Users/brianlauber/Documents/src/knex/node_modules/mocha/lib/runnable.js:382:7)
      at Runner.runTest (/Users/brianlauber/Documents/src/knex/node_modules/mocha/lib/runner.js:541:10)
      at /Users/brianlauber/Documents/src/knex/node_modules/mocha/lib/runner.js:667:12
      at next (/Users/brianlauber/Documents/src/knex/node_modules/mocha/lib/runner.js:450:14)
      at /Users/brianlauber/Documents/src/knex/node_modules/mocha/lib/runner.js:460:7
      at next (/Users/brianlauber/Documents/src/knex/node_modules/mocha/lib/runner.js:362:14)
      at Immediate.<anonymous> (/Users/brianlauber/Documents/src/knex/node_modules/mocha/lib/runner.js:428:5)
      at processImmediate (internal/timers.js:439:21) {
    name: 'KnexTimeoutError'
  }
}
{
  reason: KnexTimeoutError: Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?
      at /Users/brianlauber/Documents/src/knex/lib/client.js:351:30
  From previous event:
      at Client_MySQL2.acquireConnection (/Users/brianlauber/Documents/src/knex/lib/client.js:348:15)
      at Client_MySQL2.cancelQuery (/Users/brianlauber/Documents/src/knex/lib/dialects/mysql/index.js:175:32)
      at /Users/brianlauber/Documents/src/knex/lib/runner.js:177:37
  From previous event:
      at Runner.query (/Users/brianlauber/Documents/src/knex/lib/runner.js:169:13)
      at /Users/brianlauber/Documents/src/knex/lib/runner.js:41:23
      at /Users/brianlauber/Documents/src/knex/lib/runner.js:261:24
  From previous event:
      at Runner.ensureConnection (/Users/brianlauber/Documents/src/knex/lib/runner.js:259:8)
      at Runner.run (/Users/brianlauber/Documents/src/knex/lib/runner.js:27:12)
      at Raw.Target.then (/Users/brianlauber/Documents/src/knex/lib/interface.js:15:43)
      at Context.<anonymous> (/Users/brianlauber/Documents/src/knex/test/integration/builder/additional.js:908:10)
      at callFn (/Users/brianlauber/Documents/src/knex/node_modules/mocha/lib/runnable.js:395:21)
      at Test.Runnable.run (/Users/brianlauber/Documents/src/knex/node_modules/mocha/lib/runnable.js:382:7)
      at Runner.runTest (/Users/brianlauber/Documents/src/knex/node_modules/mocha/lib/runner.js:541:10)
      at /Users/brianlauber/Documents/src/knex/node_modules/mocha/lib/runner.js:667:12
      at next (/Users/brianlauber/Documents/src/knex/node_modules/mocha/lib/runner.js:450:14)
      at /Users/brianlauber/Documents/src/knex/node_modules/mocha/lib/runner.js:460:7
      at next (/Users/brianlauber/Documents/src/knex/node_modules/mocha/lib/runner.js:362:14)
      at Immediate.<anonymous> (/Users/brianlauber/Documents/src/knex/node_modules/mocha/lib/runner.js:428:5)
      at processImmediate (internal/timers.js:439:21) {
    name: 'KnexTimeoutError'
  }
}
```

Reply:
Q: what problem is this line intended to fix?

Reply:
Unhandled rejection Error: Transaction rejected with non-error: undefined
from https://travis-ci.org/knex/knex/jobs/650139383?utm_medium=notification&utm_source=github_status

Reply:
All troubles are going from
```js
const trx = await knex.transaction();

trx.select()
// ...blablabla...

await trx.executionPromise
```

Reply:
Gotcha.  Have you been able to recreate this error after #3666  was merged?  (The link you provided looks like it's from a Travis-CI run that is 5 days old)

Reply:
Yes, locally on node 8

Reply:
Interesting.  Are you seeing the issue locally when you run the tests?  Or, are you only seeing it when you run the snippet that you provided above?  ☝️ 

So far, I have been unable to recreate the issue when running the tests locally.  I'll give it another try once I get back to my apartment.  (in the next 30 mins or so)

Reply:
knex.transaction() - returns this.initPromise
executionPromise - rejected because of 
1) lib/transactions.js#78 - without handler executionPromise returned at line 109 and will not produce unhandled rejection
2) lib/transactions.js#198

Reply:
"rejects execution promise if there was a manual rollback and transaction is set to reject" - this test

sqlite3, node 8, several retries

Reply:
I dont understand how this supposed to work, lol

Reply:
I see first implementation and examples - no executionPromise .catch required
And then I see https://github.com/knex/knex/pull/3328

Reply:
For easy reproduction add delay after await trx.rollback()

Reply:
k -- just got back, so I'll try recreating the issue in the next few minutes.

One other thing I noticed: it appears that some test cases will "fail" if you omit a specific database from the `DB="..."` environment variable.  For instance: you will observe "failures" if you run the tests via:

```
DB="mysql mysql2" KNEX_TEST_TIMEOUT=60000 ./node_modules/.bin/mocha --exit -t 10000 test/index.js
```

However, these are not actually failures in the Knex library itself.  It's actually a problem w/ the way the test cases are written.  More specifically: the test cases are referencing data that is only available when `"sqlite3"` is specified within `DB="..."` .  (BTW: I'm going to create a separate issue for this)

So, with this in mind: what command are you running to kick off the tests?  It might provide some insight into why the tests are failing.

Reply:
@kibertoad can you help with explanation, how .executionPromise should work?
executionPromise must be awaited immediately, or after transaction commit/rollback?

Reply:
@briandamaged  
```
mocha ./test/unit/knex.js --grep "^knex rejects execution promise if there was a manual rollback and transaction is set to reject$"
```

Reply:
@maximelkin executionPromise is a way to expose state of transaction to the user, so that he can e. g. handle the rollback or proceed with the rest of the flow after commit has happened. Having that in mind, I'd say that it should be awaited after transaction commit/rollback.
Maybe we need to rename the field to make it more explicit, e. g. transactionPromise.

Reply:
@maximelkin : I've been able to recreate the failure locally using the command-line you provided.  I've identified a few suspicious sections of code, but so far none of them seem to directly impact this specific issue.

For example: take a look at the following:

https://github.com/knex/knex/blob/72886086b65583b395d2c34851ef60a5a6d5fd75/lib/transaction.js#L228-L233

These lines seem strange for at least 2 reasons:

1. There is a `.catch(()=> {})` statement, which might be hiding other bugs.  (More specifically: it might be causing Node to overlook some "unresolved promises".  This will require some more investigation to confirm this)
2. The conditional logic seems to be largely unnecessary.  In particular, `this._previousSibling` should _always_ be defined thanks to the following lines in the constructor:

https://github.com/knex/knex/blob/72886086b65583b395d2c34851ef60a5a6d5fd75/lib/transaction.js#L124-L128

But yeah: so far, these oddities don't seem to have any direct impact on the issue you're observing.

cc: @kibertoad